### PR TITLE
Reproducibility

### LINKED ISSUE #36: Random seeds seem to have no effect
When setting random seed and creating a fixed number of processes, `rf_modeling` still seems to give different results each time it's called.

### PR DESCRIPTION
**Pull request by Ben-Drucker**
See Issue #36 for details.

### CODE REVIEW

File: R/rf_modeling.R [hunk 1]
```diff
@@ -183,9 +183,15 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
         K <- nrow(dSet)
     num_rep <- ceiling(nrow(dSet)/K)
     cv_idx <- sample(rep(seq_len(K), num_rep)[seq_len(nrow(dSet))])
-    multiproc_cl <- makeCluster(max(1, detectCores() - 1))
+    if(is.null(cores)){
+        cores <- max(1, detectCores() - 1)
+    }
+    stopifnot(1 <= cores && cores <= detectCores())
```
Comment (vladpetyuk) [inline R/rf_modeling.R]: `cores <= detectCores()` either ignore this constraint or just give a warning

File: R/rf_modeling.R [hunk 2]
```diff
@@ -183,9 +183,15 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
+    multiproc_cl <- makeCluster(cores)
     clusterEvalQ(multiproc_cl, library("MSnID"))
     clusterEvalQ(multiproc_cl, library("Biobase"))
+    set.seed(42) # Set outer seed
+    clusterSetRNGStream(multiproc_cl, 21) # Set process seeds
```
Comment (vladpetyuk) [inline R/rf_modeling.R]: by conventions let's set seeds to 0. Also, let's move it into the argument list.

File: R/rf_modeling.R [hunk 3]
```diff
@@ -211,8 +230,7 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
        list(predProb, features.sel)
     }
     res <- parLapply(cl = multiproc_cl, X = 1:K, fun = fn)
-    # stopCluster(multiproc_cl) # replaced with on.exit()
-    #
+
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:240]: Formatting cleanup

File: R/rf_modeling.R [hunk 4]
```diff
@@ -175,26 +174,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
+                             envir = environment()))
     fn <- function(i){
+       RNGkind("L'Ecuyer-CMRG")
+       set.seed(i)
+       seed <- i
        i <- cv_idx == i
        if(sel.feat){
           features.sel <- FUN(x=dSet[!i,features],
-                              y=dSet[!i,response])
+                              y=dSet[!i,response], seed = seed)
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:220]: pass `seed` to downstream `FUN`

File: .lintr [hunk 5]
```diff
@@ -0,0 +1,2 @@
+linters: linters_with_defaults(cyclocomp_linter = NULL, commented_code_linter = NULL, object_name_linter = NULL, line_length_linter = line_length_linter(length = 90L))
+encoding: "UTF-8"
```
Comment (Ben-Drucker) [inline .lintr:2]: Boilerplate static analysis setup

File: R/rf_modeling.R [hunk 6]
```diff
@@ -156,7 +156,8 @@ train_model_rf <- function(x, y, ...){
 
 
 rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=TRUE,
-                            sel.alg=c("varSelRF","Boruta","top"), ...){
+                            sel.alg=c("varSelRF","Boruta","top"), cores=NULL, 
+                                seed=0, ...){
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:159]: Provided `cores` and `seed` options

File: R/rf_modeling.R [hunk 7]
```diff
@@ -175,26 +176,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
     sel.alg <- match.arg(sel.alg)
     FUN <- switch(sel.alg,
                   varSelRF = select_features_varSelRF,
-                  Boruta = function(x,y) select_features_Boruta(x,y,...),
+                  Boruta = function(x,y, ...) select_features_Boruta(x,y,...),
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:178]: Enables seed to be passed to `select_features_Boruta`.

File: R/rf_modeling.R [hunk 8]
```diff
@@ -175,26 +176,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
                   top = select_features_top)
 
     # do K-fold split here
     if(is.null(K))
         K <- nrow(dSet)
     num_rep <- ceiling(nrow(dSet)/K)
+    set.seed(seed)
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:185]: Set seed to determine `cv_idx`.

File: R/rf_modeling.R [hunk 9]
```diff
@@ -175,26 +176,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
+    })))
+    invisible(clusterExport(multiproc_cl,
                              c("dSet","cv_idx","features",
                                "response"),
-                             envir = environment())
+                             envir = environment()))
     fn <- function(i){
+       RNGkind("L'Ecuyer-CMRG")
+       set.seed(i)
+       seed <- i
```
Comment (Ben-Drucker) [inline R/rf_modeling.R]: Set random number generator kind and random seed in worker function by its index (`i`).

File: R/rf_modeling.R [hunk 10]
```diff
@@ -175,26 +176,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
+    }
+    multiproc_cl <- makeCluster(cores, outfile = "")
+
+    clusterEvalQ(multiproc_cl, invisible(suppressWarnings({
+        Sys.setenv(`_R_S3_METHOD_REGISTRATION_NOTE_OVERWRITES_` = "false")
+        suppressWarnings(suppressPackageStartupMessages({
+            library("MSnID")
+            library("Biobase")
+        }))
+    })))
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:205]: Enable quiet evaluation of library loads

File: R/rf_modeling.R [hunk 11]
```diff
@@ -175,26 +176,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
     num_rep <- ceiling(nrow(dSet)/K)
+    set.seed(seed)
     cv_idx <- sample(rep(seq_len(K), num_rep)[seq_len(nrow(dSet))])
-    multiproc_cl <- makeCluster(max(1, detectCores() - 1))
-    clusterEvalQ(multiproc_cl, library("MSnID"))
-    clusterEvalQ(multiproc_cl, library("Biobase"))
-    silence <- clusterExport(multiproc_cl,
+    if(is.null(cores)){
+        cores <- max(1, detectCores() - 1)
+    }
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:190]: Automatic # of cores determination

File: R/rf_modeling.R [hunk 12]
```diff
@@ -175,26 +176,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
+    stopifnot(1 <= cores)
+    if(cores > detectCores()){
+        msg <- sub("\n", "", "The number of specified processes is greater than
+                the number of cores available on this computer. 
+                This may lead to high computational overhead.")
+        warning(msg)
+    }
```
Comment (Ben-Drucker) [inline R/rf_modeling.R:197]: Automatic # of cores determination

File: R/rf_modeling.R [hunk 13]
```diff
@@ -175,26 +176,46 @@ rf_modeling <- function( msnset, features, response, pred.cls, K=NULL, sel.feat=
...
+    multiproc_cl <- makeCluster(cores, outfile = "")
```
Comment (Ben-Drucker) [inline R/rf_modeling.R]: Automatic # of cores determination

**Comment by Ben-Drucker** [comment]
To try out, use
```r
remotes::install_github("PNNL-Comp-Mass-Spec/MSnSet.utils", ref = "ben-reproducibility", force = TRUE)
```

**Pull request merged**